File: lsd-interceptors/src/main/java/com/nickmcdowall/lsd/http/autoconfigure/LsdFeignAutoConfiguration.java
package com.nickmcdowall.lsd.http.autoconfigure;

import com.lsd.LsdContext;
import com.nickmcdowall.lsd.http.common.HttpInteractionHandler;
import com.nickmcdowall.lsd.http.interceptor.LsdFeignLoggerInterceptor;
import com.nickmcdowall.lsd.http.naming.DestinationNameMappings;
import com.nickmcdowall.lsd.http.naming.RegexResolvingNameMapper;
import com.nickmcdowall.lsd.http.naming.SourceNameMappings;
import feign.Logger;
import lombok.RequiredArgsConstructor;
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.cloud.openfeign.FeignClientBuilder;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;
import org.springframework.context.annotation.PropertySource;

import java.util.List;

/**
 * <p>
 * If a {@link com.lsd.LsdContext} class is available it will automatically autoconfig a {@link LsdFeignLoggerInterceptor}
 * </p>
 * <br>
 * <p>
 * By default <em>source</em> name defaults to <em>'App'</em> and the <em>destination</em> name will be derived using a
 * {@link RegexResolvingNameMapper}.
 * </p>
 * <br>
 * <p>
 * Users can override either or both of the default name mappings by supplying their own {@link SourceNameMappings} or
 * {@link DestinationNameMappings} beans and naming them <em>'defaultSourceNameMapping`</em> and <em>'defaultDestinationNameMapping`</em>.
 * </p>
 */
@Configuration
@PropertySource(ignoreResourceNotFound = true, value = "classpath:lsd.properties")
@ConditionalOnProperty(name = "lsd.interceptors.autoconfig.enabled", havingValue = "true", matchIfMissing = true)
@ConditionalOnClass(value = {LsdContext.class, FeignClientBuilder.class, Logger.Level.class})
@Import({NamingConfig.class, HttpHandlerConfig.class})
@RequiredArgsConstructor
public class LsdFeignAutoConfiguration {

    private final List<HttpInteractionHandler> httpInteractionHandlers;

    @Bean
    public LsdFeignLoggerInterceptor lsdFeignLoggerInterceptor() {
        return new LsdFeignLoggerInterceptor(httpInteractionHandlers);
    }

    @Bean
    @ConditionalOnMissingBean
    public Logger.Level feignLoggerLevel() {
        return Logger.Level.BASIC;
    }
}

File: lsd-interceptors/src/main/java/io/lsdconsulting/interceptors/http/autoconfigure/LsdFeignAutoConfiguration.kt
package io.lsdconsulting.interceptors.http.autoconfigure

import com.lsd.core.LsdContext
import feign.Logger.Level
import io.lsdconsulting.interceptors.http.LsdFeignLoggerInterceptor
import io.lsdconsulting.interceptors.http.common.HttpInteractionHandler
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.cloud.openfeign.FeignClientBuilder
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Import
import org.springframework.context.annotation.PropertySource

/**
 *
 *
 * If a [LsdContext] class is available it will automatically autoconfig a [LsdFeignLoggerInterceptor]
 *
 * <br></br>
 *
 *
 * By default *source* name defaults to *'App'* and the *destination* name will be derived using a
 * [io.lsdconsulting.interceptors.http.naming.RegexResolvingNameMapper].
 *
 * <br></br>
 *
 *
 * Users can override either or both of the default name mappings by supplying their own [io.lsdconsulting.interceptors.http.naming.SourceNameMappings] or
 * [io.lsdconsulting.interceptors.http.naming.DestinationNameMappings] beans and naming them *'defaultSourceNameMapping`* and *'defaultDestinationNameMapping`*.
 *
 */
@Configuration
@PropertySource(ignoreResourceNotFound = true, value = ["classpath:lsd.properties"])
@ConditionalOnProperty(name = ["lsd.interceptors.autoconfig.enabled"], havingValue = "true", matchIfMissing = true)
@ConditionalOnClass(value = [LsdContext::class, FeignClientBuilder::class, Level::class])
@Import(NamingConfig::class, HttpHandlerConfig::class)
open class LsdFeignAutoConfiguration(
    private val httpInteractionHandlers: List<HttpInteractionHandler>
) {

    @Bean
    open fun lsdFeignLoggerInterceptor() = LsdFeignLoggerInterceptor(httpInteractionHandlers)

    @Bean
    @ConditionalOnMissingBean
    open fun feignLoggerLevel(): Level = Level.BASIC
}
